datafusion/physical-expr-common/src/sort_expr.rs

```diff
@@ -540,6 +557,33 @@ impl LexRequirement {
                 .collect(),
         )
     }
+
+    /// Constructs a duplicate-free `LexOrderingReq` by filtering out
+    /// duplicate entries that have same physical expression inside.
+    ///
+    /// For example, `vec![a Some(ASC), a Some(DESC)]` collapses to `vec![a
+    /// Some(ASC)]`.
+    ///
+    /// It will also filter out entries that are ordered if the next entry is;
+    /// for instance, `vec![floor(a) Some(ASC), a Some(ASC)]` will be collapsed to
+    /// `vec![a Some(ASC)]`.
+    pub fn collapse(self) -> Self {
+        let mut output = Vec::<PhysicalSortRequirement>::new();
+        let mut exprs = IndexSet::new();
+        let mut reqs = vec![];
+        for item in self {
+            let PhysicalSortRequirement { expr, options: req } = item;
+            // new insertion
+            if exprs.insert(expr) {
+                reqs.push(req);
+            }
+        }
+        debug_assert_eq!(reqs.len(), exprs.len());
+        for (expr, req) in izip!(exprs, reqs) {
+            output.push(PhysicalSortRequirement::new(expr, req));
+        }
+        LexRequirement::new(output)
+    }
```

Comment:
```suggestion
pub fn collapse(self) -> Self {
        let mut output = LexRequirement::default();
        for item in self {
            if !output.iter().any(|req| req.expr.eq(&item.expr)) {
                output.push(item);
            }
        }
        output
    }
```

Why doesn't this just work?

Reply:
This is a good point -- I may have been over eager to merge in the PR from @akurmustafa  -- I will think about this more carefully and either restore the original version or apply the same type of optimzization to LexOrdering::collapse

Reply:
After some more thought I reverted this change and restored it to the original version. My rationale was that I don't (yet) have evidence that this is a bottleneck for processing and thus this seems a bit like a premature optimization